drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -0,0 +1,819 @@
+
+	msg.ep = USB_CONTROL_EP_OUT;
+	msg.cb = USB_DC_EP_SETUP;
+	msg.type = USB_DC_CB_TYPE_EP;
+
+	k_msgq_put(&usb_dc_msgq, &msg, K_NO_WAIT);
+}
+
+static void udc_rpi_handle_buff_status(void)
+{
+	struct udc_rpi_ep_state *ep;
+	enum usb_dc_ep_cb_status_code status_code;
+	uint8_t status = usb_hw->buf_status;
+	unsigned int i;
+	unsigned int bit = 1U;
+	struct cb_msg msg;
+
+	LOG_DBG("status: %d", status);
+
+	for (i = 0U; status && i < USB_NUM_BIDIR_ENDPOINTS * 2; i++) {
+		if (status & bit) {
+			usb_hw_clear->buf_status = bit;
+			bool in = !(i & 1U);
+			uint8_t ep_addr = (i >> 1U) | (in ? USB_EP_DIR_IN : USB_EP_DIR_OUT);
+
+			ep = udc_rpi_get_ep_state(ep_addr);
+			status_code = in ? USB_DC_EP_DATA_IN : USB_DC_EP_DATA_OUT;
+
+			LOG_DBG("buff ep %i in? %i", (i >> 1), in);
+
+			if (i == 0 && in && state.should_set_address) {
+				state.should_set_address = false;
+				usb_hw->dev_addr_ctrl = state.addr;
+			}
+
+			if (in) {
+				k_sem_give(&ep->write_sem);
+			}
+
+			msg.ep = ep_addr;
+			msg.type = USB_DC_CB_TYPE_EP;
+			msg.cb = status_code;
+
+			k_msgq_put(&usb_dc_msgq, &msg, K_NO_WAIT);
+
+			status &= ~bit;
+		}
+
+		bit <<= 1U;
+	}
+}
+
+static void udc_rpi_isr(const void *arg)
+{
+	uint32_t status = usb_hw->ints;
+	uint32_t handled = 0;
+	struct cb_msg msg;
+
+	if (status & USB_INTS_SETUP_REQ_BITS) {
+		handled |= USB_INTS_SETUP_REQ_BITS;
+		usb_hw_clear->sie_status = USB_SIE_STATUS_SETUP_REC_BITS;
+		udc_rpi_handle_setup();
+	}
+
+	if (status & USB_INTS_BUFF_STATUS_BITS) {
+		handled |= USB_INTS_BUFF_STATUS_BITS;
+		udc_rpi_handle_buff_status();
+	}
+
+	if (status & USB_INTS_DEV_CONN_DIS_BITS) {
+		LOG_DBG("buf %u ep %u", *udc_rpi_get_ep_state(0x81)->buf_ctl,
+			*udc_rpi_get_ep_state(0x81)->ep_ctl);
+		handled |= USB_INTS_DEV_CONN_DIS_BITS;
+		usb_hw_clear->sie_status = USB_SIE_STATUS_CONNECTED_BITS;
+
+		msg.ep = 0U;
+		msg.cb = usb_hw->sie_status & USB_SIE_STATUS_CONNECTED_BITS ?
+			USB_DC_CONNECTED :
+			USB_DC_DISCONNECTED;
+		msg.type = 0U;
+
+		k_msgq_put(&usb_dc_msgq, &msg, K_NO_WAIT);
+	}
+
+	if (status & USB_INTS_BUS_RESET_BITS) {
+		int i;
+
+		LOG_WRN("BUS RESET");
+		handled |= USB_INTS_BUS_RESET_BITS;
+		usb_hw_clear->sie_status = USB_SIE_STATUS_BUS_RESET_BITS;
+		usb_hw->dev_addr_ctrl = 0;
+
+		/* The DataInCallback will never be called at this point for any pending
+		 * transactions. Reset the IN semaphores to prevent perpetual locked state.
+		 */
+
+		for (i = 0; i < USB_NUM_BIDIR_ENDPOINTS; i++) {
+			k_sem_give(&state.in_ep_state[i].write_sem);
+		}
+
+		msg.ep = 0U;
+		msg.cb = USB_DC_RESET;
+		msg.type = 0U;
+
+		k_msgq_put(&usb_dc_msgq, &msg, K_NO_WAIT);
+	}
+
+	if (status & USB_INTS_ERROR_DATA_SEQ_BITS) {
+		LOG_WRN("data seq");
+		usb_hw_clear->sie_status = USB_SIE_STATUS_DATA_SEQ_ERROR_BITS;
+		handled |= USB_INTS_ERROR_DATA_SEQ_BITS;
+	}
+
+	if (status ^ handled) {
+		LOG_ERR("unhandled IRQ: 0x%x", (uint)(status ^ handled));
+	}
+}
+
+void udc_rpi_init_bidir_endpoint(uint8_t i)
```

Comment:
Just `udc_rpi_init_endpoint(uint8_t i)`, bidir/BIDIR term is just wrong everywhere.